debian-bo-master/bo/source/net/rlpr-1.43.orig/rlpr.1
.\" $Id: rlpr.1,v 1.10 1996/11/17 08:10:55 meem Exp $
.TH rlpr 1 "1996/11/17" "printing tools -- rlpr 1.40" "UNIX Reference Manual"
.SH NAME
rlpr \- remote off line print
.SH SYNOPSIS
\fBrlpr\fP [\fB-H\fPprinthost] [\fB-P\fPprinter] [\fB-X\fPproxyhost]
[\fB-#\fPcopies] [\fB-C\fPclass] [\fB-J\fPjob] [\fB-T\fPtitle] [\fB-U\fPuser]
[\fB-i\fP[indent]] [\fB-w\fPwidth] [\fB-1234cdfghlmnopqrstvWV\fP]
[\fB--tmpdir\fP=dir]
[\fB--port\fP=port]
[\fB--verbose\fP]
[\fB--debug\fP]
[file ...]
.SH DESCRIPTION
\fBrlpr\fP uses tcp/ip to send print jobs to lpd servers anywhere on
a network.  Unlike lpr, it does not require that the remote printers
be explicitly known to the local system (traditionally through
\fI/etc/printcap\fP), and thus is considerably more flexible and
requires less administration.
.PP
\fBrlpr\fP can be used anywhere a traditional \fBlpr\fP might be used,
and is backwards compatible with traditional BSD \fBlpr\fP.
.PP
\fBrlpr\fP can be installed and used in two different ways:
.PP
1. Ideally, if the \fBrlpr\fP client can be installed setuid root,
then \fBrlpr\fP can interact directly with other \fBlpd\fP's on the
network.  See \fBexamples\fP below.
.PP
2. If \fBrlpr\fP cannot be installed setuid root (in cases where one
is not the administrator on the machine one needs to print from) then
\fBrlpr\fP can be used in conjunction with an \fBrlprd\fP proxy
agent. See the \fPconfiguring a proxy\fP and \fB examples\fP sections below
for details.
.SH MOTIVATION
Traditional Berkeley \fBlpr\fP has one major drawback: the mortal user
cannot \fBlpr\fP to a printer which is not listed explicitly in the
\fI/etc/printcap\fP file.
.PP		
This is especially annoying in scenarios where there is a central UNIX
server which many users (using their own desktop computers) connect to
in order to do their daily work; If those users wish to print files
from the central server to their personal machines (assuming they had
a local \fBlpd\fP), they \fIcannot do this\fP without the central
server's \fI/etc/printcap\fP being edited to contain the host and
printer of each user on the network \- this scales horribly.
.SH OPTIONS
.PP
.SS Environment/Setup
.TP
.I      \-\-debug
print gobs of debugging information
.TP
.I \-H, \-\-printhost=host
select the host to print to (used with \fI-P\fP).
.TP
.I \-P, \-\-printer=printername, \-\-queue=printername
select the printer to print to (used with \fI-H\fP).
.TP
.I      \-\-port=number
select an alternate port (instead of 7290) to connect to, if using
rlprd.  usually not needed.
.TP
.I \-X, \-\-proxy=proxyhost, \-\-proxyhost=proxyhost
select the proxy host to use.
.TP
.I \-\-tmpdir=tmpdir
use tmpdir for temporary files (by default, \fB/tmp\fP is used)
.TP
.I \-V, \-\-version
print version and exit.
.PP
.fi
\fBNOTE:\fP instead of constantly specifying these parameters,
one can set up defaults through either environment variables 
or \fBrlprrc\fP configuration files.  For information on personal
or system-wide \fBrlprrc\fP files, see \fBrlprrc\fP(5).
.SS Behavior
.TP
.I \-#, \-\-copies=copies
print \fIcopies\fP copies of each document \- use sparingly,
\fBprinters aren't copiers\fP.
.TP
.I \-m, \-\-mail
send mail upon completion.
.TP
.I \-r, \-\-remove
remove file after printing.
.TP
.I \-s
.br
ignored (provided for compatibility).
.TP
.I \-q, \-\-quiet, \-\-silent
quiet mode \- stay quiet (except for fatal errors).
.TP
.I \-\-verbose
verbose mode \- on by default unless \fBrlpr\fP is invoked as \fBlpr\fP.
.TP
.I \-W, \-\-windows
"Windows" mode \- for printing multiple files to a
windows-based \fBlpd\fP.
.fi
.SS Document Content
.ta \w'1234'u
.nf
\fI\-1\fP	print data using \fBtroff\fP(1) R (times roman) font.
\fI\-2\fP	print data using \fBtroff\fP(1) I (times italic) font.
\fI\-3\fP	print data using \fBtroff\fP(1) B (times bold) font.
\fI\-4\fP	print data using \fBtroff\fP(1) S (special) font.
\fI\-c\fP	assume data has been produced by \fBcifplot\fP(1).
\fI\-d\fP	assume data has been produced by \fBtex\fP(1).
\fI\-f\fP	filter data assuming the start of each line has
	a \fIfortran\fP carriage control character.
\fI\-g\fP	assume data has been produced by the BSD standard
	plot library.
\fI\-l\fP	treat control characters as regular characters.
\fI\-n\fP	assume data has been produced by \fBditroff\fP.
\fI\-o\fP	assume data is \fIpostscript\fP.
\fI\-p\fP	print data using \fBpr\fP(1).
\fI\-t\fP	assume data has been produced by \fBtroff\fP(1).
\fI\-v\fP	assume the data contains a raster image.
.fi
.SS Document Format
.TP
.I \-i, \-\-indent=[cols]
indent output by \fIcols\fP columns (8 by default).
.TP
.I \-h, \-\-no-burst
do not print burst (banner) page.
.TP
.I \-w, \-\-width=width
use \fIwidth\fP for page width when using \fBpr\fP(1).
.TP
.I \-C, \-\-class=class
use \fIclass\fP for the class name on burst page instead of
the local system name.
.TP
.I \-J,  \-\-job=jobname
use \fIjobname\fP for the job name on burst page instead of
the first filename in the job.
.TP
.I \-T, \-\-title=title
use \fItitle\fP for the title when using \fBpr\fP(1)
instead of the current filename.
.TP
.I \-U, \-\-user=username
use \fIusername\fP for the user name on burst page instead
of the actual username.
.fi
.br
.SH "CONFIGURING A PROXY"
In situations where the \fBrlpr\fP client cannot be installed setuid
root, \fBrlpr\fP cannot directly talk to an \fBlpd\fP because
\fBlpd\fP requires that print requests come from privileged
ports. Unfortunately, \fBrlpr\fP cannot "come from a privileged port"
without having superuser status.
.br
\fBThe privileged port requirement is part of RFC 1179.\fP
.PP
If you or someone you know has root access to a machine on the
network, you can use a \fBproxy\fP, known as \fBrlprd\fP to get around
this architectural wart.
.PP
\fBrlprd\fP works by taking \fBlpd\fP requests on a non-privileged port,
mapping them to a privileged port, and sending them out to the real
\fBlpd\fP. The proxy and the lpd \fBneed not be on the same
machine\fP. 
.PP
In the scenario provided in the motivation, if each user had root
access on his own machine, it would be optimal for the user to start
\fBrlprd\fP up on his own machine, in which case the proxy and the
printhost are the same machine.  The user would additionally have to
make sure his own machine was listed in his local
\fI/etc/hosts.lpd\fP so that his \fBlpd\fP would accept connections
from his \fBrlprd\fP.
.PP
However, if the machine the user wants to print to is not running
unix (but is running an \fBlpd\fP) (i.e. Windows, WinNT, etc) -- Then
that user can configure his \fBlpd\fP to accept connections from
an \fBrlprd\fP running somewhere on the network, and use that
\fBrlprd\fP as a proxy for \fBrlpr\fP.
.PP
For more on this mess, see \fBrlprd\fP(8).
.SH EXAMPLES
.TP
WITHOUT A PROXY  (rlpr is setuid root)
.PP
sun% \fBrlpr --printhost=foo.bar.org --printer=hp4l foo.c\fP
.PP
prints \fBfoo.c\fP to the printer \fBhp4l\fP on the host
\fBfoo.bar.org\fP (assuming foo.bar.org has been
configured to accept print requests from host \fBsun\fP in its
\fI/etc/hosts.lpd\fP).
.PP
sun% \fBwho | rlpr --printhost=foo.bar.org --printer=hp4l\fP
.PP
prints the output of \fBwho\fP(1) to the printer \fBhp4l\fP on the
host \fBfoo.bar.org\fP.
.TP
WITH A PROXY  (rlpr is a normal program)
.PP
sun% \fBrlpr -Xfoo.bar.org -Hfoo.bar.org -Php4l foo.c\fP
.PP
prints \fBfoo.c\fP to the printer \fBhp4l\fP on the host
\fBfoo.bar.org\fP using \fBfoo.bar.org\fP as a proxyhost. Under this
configuration, foo.bar.org needs to be configured to accept print
requests from the proxy (itself) and \fBnot from sun\fP. In situations
where the proxy and the printhost are not the same machine, this can
lead to some security holes due to the original lameness of berkeley
\fBlpd\fP's host-based security.
.PP
\fBNOTE:\fP if a proxyhost isn't specified anywhere, \fBrlpr\fP
assumes one isn't necessary.
.SH "ENVIRONMENT VARIABLES"
The following environment variables are used by \fBrlpr\fP:
.TP
.B RLPR_PRINTHOST
Specifies the default host to print to.
.TP
\fBPRINTER\fP or \fBLPDEST\fP
Specifies a default printer (printqueue) to use.
.br
First \fBPRINTER\fP is consulted, then \fBLPDEST\fP.
.TP
\fBRLPR_PROXYHOST\fP or \fBRLPR_PROXY\fP
Specifies a proxy host to use, if necessary.
.br
First \fBRLPR_PROXYHOST\fP is consulted, then \fBRLPR_PROXY\fP.
.TP
.B TMPDIR
.br
If \fBTMPDIR\fP is set, it is used for temporary files.  By default, 
\fB/tmp\fP is used.
.PP
.SH FILES
.nf
.ta \w'123456789012345678'u
$HOME/.rlprrc	Printer/Host database
$TMPDIR/cf*	Temporary control files on local machine
$TMPDIR/df*	Temporary data file for jobs from stdin
/var/spool/*/cf*	Temporary control files on printhost
/var/spool/*/df*	Temporary data files on printhost
/etc/hosts.lpd	Host-based security on printhost
/etc/passwd	Personal identification
/etc/rlprrc	System-wide Printer/Host database
.fi
.SH "SEE ALSO"
\fBrlprd\fP(8), \fBrlprrc\fP(5), \fBlpr\fP(1)
.SH AUTHOR
meem <meem@gnu.ai.mit.edu>
.SH "BUGS / LIMITATIONS"
Some evil applications have \fBlpr\fP hardcoded into them.


debian-bo-master/bo/source/net/rlpr-1.43.orig/rlpr.c
/* filename: rlpr.c
 * project: rlpr
 * author: meem  --  meem@sherilyn.wustl.edu
 * version: $Id: rlpr.c,v 1.24 1997/01/14 05:01:02 meem Exp $
 * contents: functions pertinent to the main flow of the program
 * 
 * Time-stamp: <1997/01/13 23:00 -- meem@sherilyn.wustl.edu>
 */

/* copyright (c) 1996, 1997 meem, meem@gnu.ai.mit.edu
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 1, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */

#include "config.h"

#ifndef STDC_HEADERS
#error currently there is no support for compiling on machines without ANSI headers. \
       write meem@gnu.ai.mit.edu if you are interested in adding it.
#endif

#include <assert.h>
#include <stdio.h>
#include <string.h>                   /* for strcpy() strcat(), etc */
#include <stdlib.h>                   /* for getenv(), etc */
#include <unistd.h>                   /* for unlink(), getopt(), etc */
#include <fcntl.h>                    /* for O_RDWR, etc */
#include <pwd.h>                      /* for accessing the password file */
#include <stdarg.h>

#ifdef __GNU_LIBRARY__                    
#undef __GNU_LIBRARY__
#include "getopt.h"
#define __GNU_LIBRARY__
#else  /* not __GNU_LIBRARY__ */
#include "getopt.h"
#endif /* not __GNU_LIBRARY__ */

#include "rlpr.h"
#include "rlpr-net.h"                 /* network rlpr commands are here */
#include "rlpr-dbfile.h"              /* accessing the rlpr database file */
#include "rlpr-common.h"              /* common to all rlpr c source files */
#include "rlpr-client.h"              /* common to parts of the rlpr client */

char *     name;                      /* external program name */
print_attr opts_;                     /* global printing options struct */
net_attr   net_;                      /* global network options struct  */


int main(int argc, char *argv[])
{
  int    cfd;                         /* control file descriptor */
  int    sockfd;                      /* the socket we connect to */
  char * cfname;                      /* control file name */
  char * dfname;                      /* data file name */
  char * cffullname;                  /* full pathname to control file */
  int    nf;                          /* number of files spooled */
  int    i;                           /* scrap variable */
  int    orig_argc = argc;            /* original # args */
  name = *argv;                       /* program name */

  toggle_euid();                      /* lose euid of root */

  /* this sets the opts_, props_ and net_ structures */

  argc -= parse_args(orig_argc, argv);
  argv += orig_argc - argc;

  /* create name templates for temp files */
  
  cfname = rlpr_malloc(strlen(net_.localhost) + strlen("cfAxxx") + 1);
  sprintf(cfname, "cfA%03d%s", (int) (getpid() % 1000), net_.localhost);
  dfname     = rlpr_strdup(cfname);
  dfname[0]  = 'd';
  cffullname = rlpr_malloc(strlen(cfname) + strlen(opts_.tmpdir) + 2);
  sprintf(cffullname, "%s/%s", opts_.tmpdir, cfname);

  if (!opts_.wflag) {                 /* if not connecting to a winNT lpd */
    sockfd = open_connection();       /* server now connected */
    send_recvj_req(sockfd);           /* tell lpd we have a job */
  }

  nf = argc ? argc : 1;               /* find out number of files to print */

  /*
   * open the cf once.  we're going to keep reusing the same one
   * on the client side because we only need one cf at a time
   */
  
  if ((cfd = open(cffullname, O_RDWR|O_TRUNC|O_CREAT, 0600)) < 0)
    rlpr_msg(FATAL, ERRNO, "%s - cannot create", cffullname);
  if (unlink(cffullname) < 0)
    rlpr_msg(FATAL, ERRNO, "%s - cannot unlink", cffullname);
  
  do {
    if (opts_.wflag) {                /* if connecting to a winNT lpd */
      sockfd = open_connection();
      send_recvj_req(sockfd);
    }

    lseek(cfd, 0, SEEK_SET);          /* reposition at start of file */

    cf_header(cfd, argc ? *argv : NULL); 

    for (i = 0; i < opts_.copies; i++)
      cf_add(opts_.filetype, dfname, cfd);
    
    cf_add('U', dfname, cfd);

    /* from how i read the rfc, this next 'N' line should not be necessary.
     * however, berkeley lpd's seem to require it here in order for correct
     * processing of the burst page. berkeley nonstandard? :-) never
     */

    {
      static char safe_filename[MAX_SOURCE_LEN + 1];
      cf_add('N', strncpy(safe_filename, argc ? *argv : "", MAX_SOURCE_LEN), cfd);
    }

    send_cf(sockfd, cfd, cfname);

    send_df(sockfd, argc ? *argv : NULL, dfname);
    if (opts_.rflag && argc)          /* unlink the file if user requested to */
      if (unlink(*argv) < 0)          /* and if it's possible... */
        rlpr_msg(WARNING, ERRNO, "%s - cannot unlink", *argv);
    
    argv++;
    dfname[2]++;                      /* inc. letter (dfA --> dfB), etc */
    cfname[2]++;        

    if (opts_.wflag) close_connection(sockfd); /* close sock under NT */
  } while (--argc > 0);
  
  if (!opts_.wflag) close_connection(sockfd);  /* close sock */
  
  rlpr_msg(INFO, NO_ERRNO, "%i file%s spooled to %s:%s (proxy: %s)",
           nf, (nf > 1) ? "s" : "", net_.printhost, opts_.printer,
           net_.proxyhost ? net_.proxyhost : "none");
  
  /* free() stuff.. totally unnecessary but morally satisfying */

  free(cffullname);
  free(cfname);
  free(dfname);
  
  exit(EXIT_SUCCESS);
}

/* add an entry to the control file */

void cf_add(char op, const char *str, int cfd)
{
  /* we could do an sprintf(), but we'd waste a lot of our time just
     malloc()'ing and free()'ing, so instead just employ brute force */

  safe_writen(cfd, &op, 1);
  safe_writen(cfd, str, strlen(str));
  safe_writen(cfd, "\n", 1);
}

/* uses options struct to put standard header on control file */

void cf_header(int cfd, char *filename)
{
  cf_add('H', net_.localhost, cfd);   /* hostname */
  cf_add('P', opts_.user,     cfd);   /* responsible user */

  if (opts_.indent_cols)              /* indent if appropriate */
    cf_add('I', opts_.indent_cols, cfd);

  if (opts_.mflag)                    /* mail user if appropriate */
    cf_add('M', opts_.user,   cfd);         

  if (opts_.bflag) {                  /* print burst page */

    /* the defaults on the 'J' and 'C' options are meant to appease
     * berkeley lpd, which is apparently more finicky than the RFC
     * calls for.. also the ordering of these commands mirrors
     * berkeley lpr, in case anyone is relying on the behavior
     */
    
    cf_add('J', opts_.job   ? opts_.job   : filename ? filename : "stdin", cfd);
    cf_add('C', opts_.class ? opts_.class : net_.localhost, cfd);
    cf_add('L', opts_.user,   cfd);
  }

  /* another berkeleyism that disagrees with the RFC */
  if (strcmp(opts_.width, DEFAULT_WIDTH) != 0)
    cf_add('W', opts_.width,  cfd);   /* width for f, l and p files */

  if (opts_.title)                    /* title for pr(1) */
    cf_add('T', opts_.title,  cfd);
}

int parse_args(int argc, char *argv[])
{
  int c, num_args = 1;               /* the number of arguments read */

  static struct option long_opts[] = {
    { "copies",    1, NULL, '#' },
    { "debug",     0, NULL, -6  },
    { "help",      0, NULL, -2  },
    { "indent",    1, NULL, 'i' },
    { "job",       1, NULL, 'J' },
    { "mail",      0, NULL, 'm' },
    { "no-burst",  0, NULL, 'h' },
    { "port",      1, NULL, -4  },
    { "printhost", 1, NULL, 'H' },
    { "printer",   1, NULL, 'P' },
    { "proxy",     1, NULL, 'X' },
    { "proxyhost", 1, NULL, 'X' },
    { "queue",     1, NULL, 'P' },
    { "quiet",     0, NULL, 'q' },
    { "remove",    0, NULL, 'r' },
    { "silent",    0, NULL, 'q' },
    { "tmpdir",    1, NULL, -3  },
    { "title",     1, NULL, 'T' },
    { "user",      1, NULL, 'U' },
    { "width",     1, NULL, 'w' },
    { "windows",   1, NULL, 'W' },
    { "version",   0, NULL, 'V' },
    { "verbose",   0, NULL, -5  },
    { 0, 0, 0, 0 }
  };

  init_options();                     /* initialize options struct */
  
  while ((c = getopt_long(argc, 
                          argv, 
                          "1234cdfglnopqtv#:hi::mrsw::C:FH:J:P:Q:T:U:VWX:",
                          long_opts,
                          NULL)) != EOF) {
    num_args++;
    switch(c) {

    case '1':                         /* troff type R */
    case '2':                         /* troff type I */
    case '3':                         /* troff type B */
    case '4':                         /* troff type S */
    case 'c':                         /* cifplot */
    case 'd':                         /* TeX/DVI */
    case 'f':                         /* fortran */
    case 'g':                         /* graph */
    case 'l':                         /* leave control characters */
    case 'n':                         /* ditroff */
    case 'o':                         /* postscript (not in rlpr) */
    case 'p':                         /* pr */
    case 't':                         /* troff */
    case 'v':                         /* raster input */
      opts_.filetype = (c == 'f') ? 'r' : c;
      break;
    case '#':                         /* number of copies to make */
      opts_.copies = atoi(optarg);
      break;
    case 'h':                         /* suppress printing of burst page */
      opts_.bflag  = 0;
      break;
    case 'i':                         /* indentation */
      opts_.indent_cols  = optarg ? optarg : DEFAULT_INDENT_NO_PARAM;
      break;
    case 'm':                         /* send mail after printing */
      opts_.mflag  = 1;
      break;
    case 'q':                         /* be quiet! */
      props_.quiet = 1;
      break;
    case 'r':                         /* remove file after printing */
      opts_.rflag  = 1;
      break;
    case 's':                         /* for compatibility with BSD lpr */
      rlpr_msg(WARNING, NO_ERRNO, "symlink option not applicable (ignored)");
      break;
    case 'w':                         /* width for pr */
      opts_.width  = optarg ? optarg : DEFAULT_WIDTH_NO_PARAM;
      break;
    case 'C':                         /* class */
      opts_.class  = optarg;
      break;
    case 'F':                         /* form feed after printing */
      opts_.fflag  = 1;
      break;
    case 'H':                         /* for overriding PRINTHOST */
      net_.printhost = optarg;
      break;
    case 'J':                         /* job name on burst page */
      opts_.job = optarg;
      break;
    case 'P':                         /* printer queue name (FALLTHRU) */
    case 'Q':                         /* i think this is a better name */
      opts_.printer = optarg;
      break;
    case 'T':                         /* title for pr (def: filename) */
      opts_.title = optarg;
      break;
    case 'U':                         /* username to print on title... */
      opts_.user = optarg;
      break;
    case 'V':                         /* print version info */
      fprintf(stdout, "%s: version "VERSION" from "__DATE__" "__TIME__ 
              " -- meem@gnu.ai.mit.edu\n", name);
      exit(EXIT_SUCCESS);
    case 'W':
      opts_.wflag = 1;
      break;
    case 'X':                         /* for overriding PROXY  */
      net_.proxyhost = optarg;
      break;
    case -2:
      printf("usage: rlpr [-Hprinthost]  [-Pprinter]  [-Xproxyhost] [OPTIONS] "
             "[file ...]\n\nplease see the manpage for detailed help.\n");
      exit(EXIT_SUCCESS);
    case -3:
      opts_.tmpdir = optarg;          /* tmpdir to use */
      break;
    case -4:
      net_.port = atoi(optarg);       /* port number to connect to */
      break;
    case -5:                          /* verbose flag */
      props_.quiet = 0;
      break;
    case -6:
      props_.debug = 1;
      break;
    case '?':
      fprintf(stderr,"please see manpage for help\n");
      break;
    }
  }
  tidy_and_check_options();
  return num_args;
}


void init_options(void)
{
  char *tmp;

  opts_.fflag    = 0;                 /* don't auto-form feed by default */
  opts_.bflag    = 1;                 /* burst page on by default */
  opts_.mflag    = 0;                 /* don't mail after printing */
  opts_.rflag    = 0;                 /* don't remove after printing */
  opts_.wflag    = 0;                 /* windows braindead mode (off) */
  opts_.copies   = 1;                 /* number of copies to make */
  opts_.job      = NULL;
  opts_.class    = NULL;
  opts_.title    = NULL;
  opts_.filetype = 'f';               /* default to a regular file */
  opts_.width    = DEFAULT_WIDTH;
  opts_.indent_cols = NULL;           /* number of columns to indent output */

  { /* find out what user we are -- user can always provide an alternate
     * choice with the -u option
     */

    struct passwd *pw;
    if ((pw = getpwuid(getuid())) != NULL)
      opts_.user = rlpr_strdup(pw->pw_name);
    else rlpr_msg(FATAL, ERRNO, "who are you?");
  }
  
  /* be quiet by default if we were called as "lpr" */
  props_.quiet   = strcmp(name, "lpr") ? 0 : 1;
  props_.debug   = 0;                 /* debugging off by default */
  props_.syslog  = 0;

  tmp = getenv("TMPDIR");
  opts_.tmpdir   = tmp ? tmp : DEFAULT_TMP_DIR;
  
  /* first consult PRINTER, if it's NULL then LPDEST */
  tmp = getenv("PRINTER");
  opts_.printer  = tmp ? tmp : getenv("LPDEST");        

  /* configure net_ stuff */
  net_.port      = DEFAULT_RLPRD_TO_PORT;
  tmp = getenv(PROXYHOST);
  net_.proxyhost = tmp ? tmp : getenv(PROXYHOST2);

  net_.printhost = getenv(PRINTHOST);
  net_.localhost = rlpr_malloc(MAXHOSTNAMELEN);
  if (get_local_hostname(net_.localhost, MAXHOSTNAMELEN) == -1)
    rlpr_msg(FATAL, NO_ERRNO, "unable to resolve your local hostname!");
}

void tidy_and_check_options(void)
{
  /* resolve remainder of printer/printhost information */
  
  if (net_.printhost == NULL && opts_.printer == NULL)
    rlpr_msg(FATAL, NO_ERRNO, "no printhost or printer set - read %s(1)", name);

  if (net_.printhost == NULL)
    net_.printhost = gethostfromq(opts_.printer);
  else if (opts_.printer == NULL)
    opts_.printer  = getqfromhost(net_.printhost);

  /* now we need to go through all the options and make sure they're
   * not too large.. if they are, we may overflow the limits of some
   * poorly-coded lpd's
   */

  rlpr_msg(DEBUG, NO_ERRNO, "checking user-supplied params...");
  if (opts_.printer)  assert(strlen(opts_.printer)  < MAX_QUEUE_LEN);
  if (opts_.title)    assert(strlen(opts_.title)    < MAX_TITLE_LEN);
  if (opts_.job)      assert(strlen(opts_.job)      < MAX_JOB_LEN);
  if (opts_.user)     assert(strlen(opts_.user)     < MAX_USER_LEN);
  if (opts_.class)    assert(strlen(opts_.class)    < MAX_CLASS_LEN);
  if (net_.localhost) assert(strlen(net_.localhost) < MAXHOSTNAMELEN);
  if (net_.printhost) assert(strlen(net_.printhost) < MAX_HOST_LEN);
  if (net_.proxyhost) assert(strlen(net_.proxyhost) < MAXHOSTNAMELEN);
  if (opts_.width)    assert(strlen(opts_.width)    < MAX_STR_LEN);
  if (opts_.indent_cols) assert(strlen(opts_.indent_cols) < MAX_STR_LEN);
  rlpr_msg(DEBUG, NO_ERRNO, "passed");
}

